### PR TITLE
NEW DataQuery::exists now generates EXISTS SQL statements

### PR DESCRIPTION
**Pull request by ScopeyNZ**
This PR adds an `exists` method to DataQuery that will generate an `EXISTS` statement from the existing query. This basically looks like this:

```
SELECT EXISTS(SELECT * FROM MyDataObject WHERE SomeProperty)
```

This is the preferred way to do these sorts of queries as the engine can just do checks like "does the `MyDataObject.SomeProperty` index have _any_ entries for `true`" - this can be a lot faster than counting results.

### CODE REVIEW

**Comment by ScopeyNZ** [comment]
I just realised I forgot to ensure tests are passing 🤦‍♂️ . I'll hope travis has good news.

**Comment by ScopeyNZ** [comment]
Ok the remaining failing tests are fixed by @sminnee's recent updates to consistent boolean return types and relies on a fix in 2.x of the postgres module.

Should I update travis config or update my code to be compatible with older versions of postgres?

**Comment by robbieaverill** [comment]
>Should I update travis config or update my code to be compatible with older versions of postgres?

It sounds like that should be done upstream, but you could patch it here too if you want to

File: src/ORM/DataQuery.php [hunk 1]
```diff
@@ -447,6 +448,33 @@ public function count()
...
+     * performant - especially when used in combination with indexed columns (that you're filtering on)
+     *
+     * @return bool
+     */
+    public function exists()
+    {
+        // Grab a statement selecting "everything" - the engine shouldn't care what's being selected in an "EXISTS"
+        // statement anyway
+        $statement = $this->getFinalisedQuery(null, true);
+        $statement->setSelect('*');
```
Comment (michalkleiner) [inline src/ORM/DataQuery.php]: What about `$statement->setSelect('1');` ? If we're only checking existence, we don't need to select all columns.

Comment (ScopeyNZ) [reply src/ORM/DataQuery.php]: Good question. I'm almost certain that the query optimiser treats `*` and `1` the same when within an `EXISTS` query. The `*` is purely SQL semantics.

File: src/ORM/DataQueryManipulator.php [hunk 2]
```diff
@@ -15,8 +15,14 @@ interface DataQueryManipulator
...
      * @param SQLSelect $sqlSelect
+     * @param bool $preferConditions
      */
-    public function beforeGetFinalisedQuery(DataQuery $dataQuery, $queriedColumns = [], SQLSelect $sqlSelect);
+    public function beforeGetFinalisedQuery(
+        DataQuery $dataQuery,
+        $queriedColumns = [],
+        SQLSelect $sqlSelect,
+        $preferConditions = false
+    );
```
Comment (kinglozzer) [inline src/ORM/DataQueryManipulator.php]: Isn’t this a breaking API change? :(

File: src/ORM/DataQuery.php [hunk 3]
```diff
@@ -447,6 +448,34 @@ public function count()
...
+        $statement->setGroupBy(null);
+        $statement->setHaving(null);
+
+        // Wrap the whole thing in an "EXISTS"
+        $sql = 'SELECT EXISTS(' . $statement->sql($params) . ')';
+        $result = DB::prepared_query($sql, $params);
+        $row = $result->first();
+        $result = reset($row);
+
+        return $result === 1 || $result === 't';
```
Comment (kinglozzer) [inline src/ORM/DataQuery.php]: Commit history tells me that this is because PGSQL - if we can’t use `(bool)$result` or similar, could you add a comment explaining what `'t'` represents here?

File: src/ORM/FieldType/DBForeignKey.php [hunk 4]
```diff
@@ -84,12 +84,21 @@ public function scaffoldFormField($title = null, $params = null)
...
+            // Let the DB do the threshold check as it will be faster - depending on the SQL engine it might only have
+            // to count indices
+            $dataQuery = $list->dataQuery()->getFinalisedQuery();
+            $dataQuery->setDistinct(false);
+            $dataQuery->setOrderBy(false);
+            $dataQuery->setSelect(['over_threshold' => 'count(*) > ' . (int) $threshold]);
+            $result = $dataQuery->execute()->column('over_threshold')[0];
             static::$foreignListCache[$hasOneClass] = [
-                'count' => $list->count(),
+                'overThreshold' => is_bool($result) ? $result : ($result === '1' || $result === 't'),
```
Comment (kinglozzer) [inline src/ORM/FieldType/DBForeignKey.php]: Could you break these changes out into a separate commit?

**Comment by kinglozzer** [comment]
Looks like a great addition! If it’s possible to work around the BC issue I’d be happy to take this in

File: src/ORM/DataQueryManipulator.php (hunk 2 above)
Comment (ScopeyNZ) [reply src/ORM/DataQueryManipulator.php]: Yeah good point... I didn't consider that. I'll think about this a little more...

File: src/ORM/DataQuery.php (hunk 3 above)
Comment (ScopeyNZ) [reply src/ORM/DataQuery.php]: The PG adapter returns `'t'`'s and `'f'`'s for true/false. Although the tests are still failing so I might have made a mistake...

Comment (sminnee) [reply src/ORM/DataQuery.php]: Note that I’ve fixed this but it probably requires another minor release of the pgsql module.

Comment (ScopeyNZ) [reply src/ORM/DataQuery.php]: Yeah I think I'll just return to using `$result === 1` and ignore PostgreSQL. I can update the travis builds to use the new release of the postgres module.

File: src/ORM/DataQueryManipulator.php (hunk 2 above)
Comment (ScopeyNZ) [reply src/ORM/DataQueryManipulator.php]: I might move `preferConditions` to a property of the query. Then you can do something like `setPreferConditions(bool $preference);`.

File: src/ORM/DataQuery.php [hunk 5]
```diff
@@ -447,6 +448,34 @@ public function count()
...
+        // Grab a statement selecting "everything" - the engine shouldn't care what's being selected in an "EXISTS"
+        // statement anyway
+        $statement = $this->getFinalisedQuery(null, true);
+        $statement->setSelect('*');
+
+        // Clear distinct, grouping, order, and having as it's not relevant for an exists query
+        $statement->setDistinct(false);
+        $statement->setOrderBy(null);
+        $statement->setGroupBy(null);
+        $statement->setHaving(null);
```
Comment (tractorcow) [inline src/ORM/DataQuery.php]: Having and group by absolutely will be necessary, as it will affect whether exists() returns rows or not for matching groups.

Comment (tractorcow) [reply src/ORM/DataQuery.php]: And with grouping, you need to respect the original select columns don't you? :P

File: src/ORM/DataQuery.php [hunk 6]
```diff
@@ -196,9 +196,10 @@ public function setQueriedColumns($queriedColumns)
      * Ensure that the query is ready to execute.
      *
      * @param array|null $queriedColumns Any columns to filter the query by
+     * @param bool $preferConditions Indicates that conditions should be preferred over joins (used with "EXISTS")
      * @return SQLSelect The finalised sql query
      */
-    public function getFinalisedQuery($queriedColumns = null)
+    public function getFinalisedQuery($queriedColumns = null, $preferConditions = false)
```
Comment (tractorcow) [inline src/ORM/DataQuery.php]: I may be a bit blind, but I see $preferConditions arg added, but I don't see any logic that depends on it. Is there some other part of the code that respects this value?

Comment (tractorcow) [reply src/ORM/DataQuery.php]: If core ignores this value, could I argue it shouldn't be necessary?

Comment (ScopeyNZ) [reply src/ORM/DataQuery.php]: It's given to `augmentSQL`. I think it's reasonable that it can be added to core functionality too but I can't think of a use-case. 

It's difficult to gain detail about the query from extensions when using `limit` or `exists` so this is an indicator of that. Perhaps `preferConditions` is the wrong name.

File: src/ORM/DataQuery.php (hunk 5 above)
Comment (tractorcow) [reply src/ORM/DataQuery.php]: E.g. `SELECT "Value", COUNT("Value") as "CountValue" FROM "Table" GROUP BY "Value" HAVING "CountValue" > 1` should return false for exists if no item has more than one value, but true as soon as a value is in the table twice.

Comment (ScopeyNZ) [reply src/ORM/DataQuery.php]: With or without the group there'll be at least one record right? Having is necessary though. I think that's a mistake.

Comment (ScopeyNZ) [reply src/ORM/DataQuery.php]: Right, if we keep the having then the group is relevant

Comment (ScopeyNZ) [reply src/ORM/DataQuery.php]: I don't think the original columns ever matter though - records are not filtered out based on what you select - I think. You either have records or you don't. Respecting `count(*), ParentID` or something when `GROUP BY ParentID` doesn't matter because you either have rows or you don't.

Comment (ScopeyNZ) [reply src/ORM/DataQuery.php]: > Right, if we keep the having then the group is relevant

I'm confused - the having doesn't rely on the group. So I'm back to the group isn't required 😆

**Comment by ScopeyNZ** [comment]
Ok I've updated this PR to only contain the `EXISTS` query changes. I'll raise additional PRs for the threshold fix and the idea about the "prefer conditions" API.

File: src/ORM/DataQuery.php (hunk 3 above)
Comment (ScopeyNZ) [reply src/ORM/DataQuery.php]: I've added a comment. While this is redundant in PostgreSQL 2.2 it's probably no harm to have a little BC for the Postgres users.

File: src/ORM/DataQueryManipulator.php (hunk 2 above)
Comment (ScopeyNZ) [reply src/ORM/DataQueryManipulator.php]: Resolving the conversation here but I'll link to it in the new PR I'll make soon

File: src/ORM/FieldType/DBForeignKey.php (hunk 4 above)
Comment (ScopeyNZ) [reply src/ORM/FieldType/DBForeignKey.php]: I'm going to raise this as a patch PR separately.

File: src/ORM/DataQuery.php (hunk 5 above)
Comment (ScopeyNZ) [reply src/ORM/DataQuery.php]: I've removed the `having`. I did a little research to back up my thoughts that `GROUP BY` remains superfluous in these `EXISTS` queries - and I'm pretty sure it is. I can't think of any scenario where adding a `GROUP BY` will change the returned records from "some" to "none".

File: src/ORM/DataQuery.php (hunk 6 above)
Comment (ScopeyNZ) [reply src/ORM/DataQuery.php]: I'm moving this change to a new PR. I'll reference this feedback in it when I raise it.

File: src/ORM/DataQueryManipulator.php (hunk 2 above)
Comment (ScopeyNZ) [reply src/ORM/DataQueryManipulator.php]: I've decided to drop this idea. It's probably complexity that we don't need.

File: src/ORM/DataQuery.php (hunk 6 above)
Comment (ScopeyNZ) [reply src/ORM/DataQuery.php]: I've decided to drop it - thanks for prompting me to think this through a little more 😅

**Comment by ScopeyNZ** [comment]
Hmmmn.. Tests...

image

File: src/ORM/DataQuery.php (hunk 5 above)
Comment (tractorcow) [reply src/ORM/DataQuery.php]: How does your having work if you don't group anymore then?

Comment (ScopeyNZ) [reply src/ORM/DataQuery.php]: Aggregates in `HAVING` don't require a `GROUP BY`?

Comment (ScopeyNZ) [reply src/ORM/DataQuery.php]: I googled this to be double sure I wasn't making a bad assumption - here's the stack exchange result I got 😂 :

https://dba.stackexchange.com/questions/57445/use-of-having-without-group-by-in-sql-queries

Comment (ScopeyNZ) [reply src/ORM/DataQuery.php]: Ok this _can_ break if you're doing your own aggregates an aliasing I guess

**Comment by ScopeyNZ** [comment]
I figured out the failing tests: https://github.com/silverstripe/silverstripe-framework/pull/8934

File: src/ORM/DataQuery.php [hunk 7]
```diff
@@ -447,6 +447,40 @@ public function count()
...
+        // statement anyway
+        $statement = $this->getFinalisedQuery();
+        $statement->setSelect('*');
+
+        // Clear limit, distinct, and order as it's not relevant for an exists query
+        $statement->setDistinct(false);
+        $statement->setOrderBy(null);
+        $statement->setLimit(null);
+
+        // We can remove grouping if there's no "having" that might be relying on an aggregate
```
Comment (tractorcow) [inline src/ORM/DataQuery.php]: Ah yes that's the solution I was after. :) good job.

File: src/ORM/DataQuery.php [hunk 8]
```diff
@@ -447,6 +447,40 @@ public function count()
...
+     * performant - especially when used in combination with indexed columns (that you're filtering on)
+     *
+     * @return bool
+     */
+    public function exists()
+    {
+        // Grab a statement selecting "everything" - the engine shouldn't care what's being selected in an "EXISTS"
+        // statement anyway
+        $statement = $this->getFinalisedQuery();
+        $statement->setSelect('*');
```
Comment (tractorcow) [inline src/ORM/DataQuery.php]: This MIGHT have to go inside the below `if (empty($having)) {` condition.

I suggest adding some more tests for grouping and having queries having ->exists() calls on them just to be sure. :)

Comment (ScopeyNZ) [reply src/ORM/DataQuery.php]: Good call. I really want to get this in to 4.4 so I'll follow up on this on Friday

Comment (ScopeyNZ) [reply src/ORM/DataQuery.php]: Ok I looked into this a bit. You're right that it "MIGHT". It's not part of the SQL standard to allow `HAVING` clauses using aliased columns. It doesn't work at all on PostgreSQL. MySQL lets you do it no problem though. I'll move it in for MySQL.

**Comment by ScopeyNZ** [comment]
I added a very simple test. I tried doing some craziness with aggregates and `HAVING` but the API actually makes it really hard to do as you can't really select the field you're aggregating in your `HAVING` clause and that's not so easy to with the `DataList` API.

**Comment by emteknetnz** [comment]
This pull request hasn't had any activity for a while. Are you going to be doing further work on it, or would you prefer to close it now?

**Comment by robbieaverill** [comment]
@emteknetnz this isn't waiting for the author

**Comment by sminnee** [comment]
I resolved the conflict but I'd probably give such an old branch a rebase before merging.

**Comment by sminnee** [comment]
...rebased.

**Review by sminnee — APPROVED** [review]
If the tests pass I'm happy with this.

File: src/ORM/DataQuery.php [hunk 9]
```diff
@@ -453,6 +453,41 @@ public function count()
         return $this->getFinalisedQuery()->count("DISTINCT {$quotedColumn}");
     }
 
+    /**
+     * Return whether this dataquery will have records. This will use `EXISTS` statements in SQL which are more
+     * performant - especially when used in combination with indexed columns (that you're filtering on)
+     *
+     * @return bool
+     */
+    public function exists()
```
Comment (robbieaverill) [inline src/ORM/DataQuery.php:462]:
```suggestion
    public function exists(): bool
```

May as well since it's a new API
